### ISSUE TITLE
Document `NXF_ENABLE_CACHE_INVALIDATION_ON_TASK_DIRECTIVE_CHANGE`

### ISSUE THREAD

**Issue by ewels** [nextflow-io/nextflow#4326]
Nextflow version 21.12.0-edge introduced changes to how cache works. A new env var was added to turn this behaviour off, called `NXF_ENABLE_CACHE_INVALIDATION_ON_TASK_DIRECTIVE_CHANGE`. However, it is not documented anywhere.

More details:
- https://github.com/nextflow-io/nextflow/issues/2382

**Comment by pditommaso** [comment]
Because it was meant to have a way to revert the previous behaviour if some unforeseen behaviour were introduced. Think this setting should be removed

**Comment by ewels** [comment]
Yeah, I had a feeling you might say that. Context for where this issue came from is that this particular bug was actually a feature for someone. In this case, `NXF_ENABLE_CACHE_INVALIDATION_ON_TASK_DIRECTIVE_CHANGE` was the best solution (I think), and it's a hidden / undocumented option.

If the intention is to remove it again then I agree that we shouldn't document it.

If the context below is convincing, then we could perhaps change the plan and pivot to keeping it as a desired feature flag.

## Context

Changing `process.ext.prefix` and resuming _should_ break the cache, as reported and changed in #2382. That was a good change in the majority of cases ✅ 

However, in some cases, changes to the task config may not be so relevant. What came up [in Slack](https://nextflow.slack.com/archives/C02T98A23U7/p1694636732261389) was the following:

* Run a process with 1000 tasks - 1 fails because not enough CPUs
* Update the config to give move CPUs, run with resume to get that one process to complete
* Cache is invalidated, all 100 tasks have to run from scratch again

> Well, let's say I'm trying to assemble reads with metaSPAdes (where you can specify memory), and my pipeline is set up so that it works with 128.GB memory. Now I'm processing some samples (say 1,000) and I didnt expect a few of them to be so large. So after the pipeline finishes, I have a few failed samples, and I decide to up my resource requirements to finish it. Unfortunately here, I won't be able to resume
>
> That being said, dynamic retry works as a solution here. It's just that in some cases, if you couldnt foresee the range of resources you needed ahead of time, you wont be able to change it

@stevekm then wrote up this very nice comparison of behaviours: https://github.com/stevekm/nextflow-demos/tree/master/dynamic-directive-per-sample

Here - on a _reproducibility_ basis, invalidating the cache is the right thing to do - albeit very strict behaviour.

In a _practical_ sense, being able to relax the caching strategy on a single per-run basis, just to avoid having to re-run a tonne of compute, when the end user is personally responsible for the reproducibility aspect, having this env var available is a good thing.

**Comment by pditommaso** [comment]
How ext.prefix is related to CPUs?

**Comment by ewels** [comment]
It's not, but both are process directives.

* `ext.prefix` is used to change the run time command - resuming is bad, the current default behaviour is good.
* `task.cpus` should have mostly non-destructive changes to the run time - resuming may be ok, if explicitly chosen

**Comment by pditommaso** [comment]
Indeed, several users complained that modifying the `cpus` and `memory` directives result in the cache invalidation.

I'm starting to think the "correct" behavior was the legacy, ie. directives changes should not invalid the cache. However, there's the case  of the `ext.something` which is widely used to customise the task script, for which the invalidation logic has been changed. 

Moreover looking into the nf-core there are essentially 4 `task.` directives accessed by the pipeline code 

- task.ext.something 
- task.cpus 
- task.memory 
- task.process (which is immutable by definition)

I think we should rollback the previous logic (no invalidation on directive change) with the exception of the `task.ext.` directives that will be considered a special one, for his extension nature. 

Thoughts?

**Comment by bentsherman** [comment]
I always assumed that the rendered task script was used for the task hash, but I recently discovered that only the script *source* is used in combination with task variables.

I think it would be better if we simply use the rendered task script and not include and variables or directives. That way, both directives and task variables would affect the cache only insofar as they are used in the task script.

**Comment by ewels** [comment]
Yes I always assumed the same until the conversation thread happened that lead to this issue!

The only problem with using the rendered script is that cpus and memory are often used in commands. So for a certain fraction of cases (including the OP above), that would be no better.

Paolo's suggestion would be good though 👍🏻

**Comment by pditommaso** [comment]
I think originally it was used the rendered script, but it was changed to the task source.

**Comment by bentsherman** [comment]
Resolved by #4339
